pkg/tools/serialize/serialize.go

```diff
@@ -19,59 +19,48 @@
 // will be executed in order
 package serialize
 
-import "runtime"
+import (
+	"sync"
+	"sync/atomic"
+)
 
-// Executor - allows one at a time in order execution of func()s.  func()s can be queued Asynchronously or
-// Synchronously
-type Executor interface {
-	AsyncExec(func())
-	SyncExec(func())
-}
+const (
+	channelSize = 100
+)
 
-type executor struct {
-	execCh      chan func()
-	finalizedCh chan struct{}
+// Executor - a struct that can be used to guarantee exclusive, in order execution of functions.
+type Executor struct {
+	execCh  chan func()
+	init    sync.Once
+	drained uint32
 }
 
-// NewExecutor - returns a new Executor
-func NewExecutor() Executor {
-	rv := &executor{
-		execCh:      make(chan func(), 100),
-		finalizedCh: make(chan struct{}),
-	}
-	go eventLoop(rv.execCh, rv.finalizedCh)
-	runtime.SetFinalizer(rv, func(f *executor) {
-		close(f.finalizedCh)
+// Exec - guarantees f() will be executed Exclusively and in the Order submitted.
+//        It immediately returns a channel that will be closed when f() has completed execution.
+func (e *Executor) Exec(f func()) (done chan struct{}) {
```

Comment:
I would prefer to use some kind of Submit or leave AsyncExec name, Since Exec is not looks like async one.

Reply:
After some consideration, have decided you are correct, also decided to preserve backward compatibility with NewExecutor().  That simplifies the PR a lot.